lib/compat/wordpress-6.6/blocks.php

```diff
@@ -44,3 +44,237 @@ function gutenberg_replace_pattern_override_default_binding( $parsed_block ) {
+			if ( ! empty( $closer_tag_bookmark->end ) ) {
+				$this->lexical_updates[] = new WP_HTML_Text_Replacement( $closer_tag_bookmark->end + 1, $closer_tag_bookmark->end + 1, $new_element );
+				return true;
+			}
+			$after_closer_tag = $closer_tag_bookmark->start + $closer_tag_bookmark->length;
+			/*
+			 * There was a bug in the HTML Processor token length fixed after 6.5.
+			 * This check is needed to add compatibility for that.
+			 * Related issue: https://github.com/WordPress/wordpress-develop/pull/6625
+			 */
+			if ( '>' === $this->html[ $after_closer_tag ] ) {
+				++$after_closer_tag;
+			}
+
+			// Append the new element.
+			$this->lexical_updates[] = new WP_HTML_Text_Replacement( $after_closer_tag, 0, $new_element );
+			return true;
+		}
+
+		/**
+		 * Remove the current figcaption tag element.
+		 *
+		 * @return bool Whether the element was properly removed.
+		 */
+		public function gutenberg_remove_figcaption_tag_element() {
+			$tag_name = $this->get_tag();
+			// Ensure only figcaption is removed.
+			if ( 'FIGCAPTION' !== $tag_name ) {
+				return false;
+			}
+			// Set position of the opener tag.
+			$this->set_bookmark( 'opener_tag' );
+
+			// Visit the closing tag.
+			if ( ! $this->next_tag(
+				array(
+					'tag_name'    => $tag_name,
+					'tag_closers' => 'visit',
+				)
+			) || ! $this->is_tag_closer() ) {
+				return false;
+			}
+
+			// Set position of the closer tag.
+			$this->set_bookmark( 'closer_tag' );
+
+			// Get position of the tags.
+			$opener_tag_bookmark = $this->bookmarks['opener_tag'];
+			$closer_tag_bookmark = $this->bookmarks['closer_tag'];
+
+			// Compat for 6.4, where bookmarks and WP_HTML_Text_Replacement are different.
+			if ( ! empty( $closer_tag_bookmark->end ) ) {
+				$this->lexical_updates[] = new WP_HTML_Text_Replacement( $opener_tag_bookmark->start, $closer_tag_bookmark->end + 1, '' );
+				return true;
+			}
+
+			$after_closer_tag = $closer_tag_bookmark->start + $closer_tag_bookmark->length;
+			/*
+			 * There was a bug in the HTML Processor token length fixed after 6.5.
+			 * This check is needed to add compatibility for that.
+			 * Related issue: https://github.com/WordPress/wordpress-develop/pull/6625
+			 */
+			if ( '>' === $this->html[ $after_closer_tag ] ) {
+				++$after_closer_tag;
+			}
+			$current_tag_length = $after_closer_tag - $opener_tag_bookmark->start;
+
+			// Remove the current tag.
+			$this->lexical_updates[] = new WP_HTML_Text_Replacement( $opener_tag_bookmark->start, $current_tag_length, '' );
+			return true;
+		}
+	};
+
+	/*
+	 * For backward compatibility, the logic from the image render needs to be replicated.
+	 * This is because the block attributes can't be modified with the binding value with `render_block_data` filter,
+	 * as it doesn't have access to the block instance.
+	 */
+	if ( $block_reader->next_tag( 'figure' ) ) {
+		$block_reader->set_bookmark( 'figure' );
+	}
+
+		$new_value = wp_kses_post( $source_value );
```

Comment:
Maybe we can move this escaping to `gutenberg_set_content_between_figcaption_balanced_tags` and `gutenberg_append_figcaption_element_after_tag`.
It will be repeated in two functions, but we will follow the [escape late approach](https://developer.wordpress.org/apis/security/escaping/#toc_3).